app/services/foreman/provision/ssh.rb

```diff
@@ -50,64 +69,67 @@ def command_prefix
-    # tell if other locations are writeable or executable by the user.
+  # Use the users home to store the provision script since we can't reliably
+  # tell if other locations are writeable or executable by the user.
+  def main_command
     main_execution = "(chmod 0701 ./#{remote_script} && #{command_prefix} ./#{remote_script} ; echo $? >#{remote_script}.status) 2>&1"
     "#{command_prefix} sh -c '#{main_execution} | tee #{remote_script}.log; exit $(cat #{remote_script}.status)'"
   end
 
-  def defaults
-    {
-      :keys_only    => true,
-      :config       => false,
-      :auth_methods => %w(publickey),
-      :compression  => true,
-      :logger       => logger,
-    }
-  end
-
   def logger
     Rails.logger
   end
 
-  def initiate_connection!
-    Timeout.timeout(Setting[:ssh_timeout].to_i) do
-      Timeout.timeout(8) do
-        ssh.run('pwd')
-      end
-    rescue Errno::ECONNREFUSED
-      logger.debug "Connection refused for #{address}, retrying"
-      sleep(2)
-      retry
-    rescue Errno::EHOSTUNREACH
-      logger.debug "Host unreachable for #{address}, retrying"
-      sleep(2)
-      retry
-    rescue Net::SSH::Disconnect
-      logger.debug "Host dropping connections for #{address}, retrying"
-      sleep(2)
-      retry
-    rescue Net::SSH::ConnectionTimeout
-      logger.debug "Host timed out for #{address}, retrying"
-      sleep(2)
-      retry
-    rescue Net::SSH::AuthenticationFailed
-      logger.debug "Auth failed for #{username} at #{address}, retrying"
-      sleep(2)
-      retry
-    rescue Timeout::Error
-      retry
-    rescue => e
-      Foreman::Logging.exception("SSH error", e)
-    end
+  def scp
+    logger.debug "about to upload #{template} to remote system at #{remote_script}"
+    logger.debug scp_cmd
+    _stdout, stderr, status = Open3.capture3(*scp_cmd)
+
+    return if status.success?
+
+    logger.error stderr
+    raise Foreman::Exception.new("Failed to upload '#{@template}' to remote system at #{address}")
   end
 
-  def ssh
-    Fog::SSH.new(address, username, options.merge(:timeout => 4))
+  def passwd_cmd
+    return if @options[:password].empty?
+    ["sshpass", "-p", @options[:password]]
   end
 
-  def scp
-    Fog::SCP.new(address, username, options)
+  def ssh_key
+    return if @options[:key_data].empty?
+    ["-i", @tmp_key_path]
+  end
+
+  def ssh_cmd(host_command)
+    [passwd_cmd, "ssh", ssh_key, ssh_options, remote_host, host_command].flatten.compact
+  end
+
+  def scp_cmd
+    [passwd_cmd, "scp", ssh_key, ssh_options, @template, "#{remote_host}:~/#{remote_script}"].flatten.compact
+  end
+
+  def ssh_options
+    opts = ["ConnectTimeout=6", "StrictHostKeyChecking=accept-new",
+            "Compression=yes"]
+
+    if @options[:key_data].present?
+      opts << "PubkeyAuthentication=yes"
+      opts << "PasswordAuthentication=no"
+    else
+      opts << "PasswordAuthentication=yes"
+      opts << "PubkeyAuthentication=no"
+    end
+
```

Comment:
Originally I had something like this in mind, but what you have here should work if we operate under the assumption that we'll have either the key or the password, but not both at once. And even then we might be fine with the key just winning.

```suggestion
    methods = []
    methods << 'password' if @options[:password].present?
    methods << 'publickey' if @options[:key_data].present?
    raise "Can't use any authentication method" if methods.empty?

    opts << "PreferredAuthentications=#{methods.join(',')}"

```